### PR TITLE
Return errors rather than assert

### PR DESCRIPTION
**Pull request by xoloki**
When computing secrets or making aggregate signatures, we previously would ```assert``` if we encountered any errors.  This change instead returns errors from ```Party``` and ```SignatureAggregator```.

### CODE REVIEW

File: src/v1.rs [hunk 1]
```diff
@@ -92,22 +93,54 @@ impl Party {
...
+        }
+
         self.private_key = Scalar::zero();
+        let mut bad_ids = Vec::new();
+        for (i, _s) in shares.iter() {
+            let Ai = &A[*i];
+            if !Ai.verify() {
+                bad_ids.push(*i);
+            }
+        }
```
Comment (netrome) [inline src/v1.rs:119]:
```suggestion
        let bad_ids = shares.keys().cloned().filter(|i| A[*i].verify()).collect();
```

File: src/v1.rs [hunk 2]
```diff
@@ -92,22 +93,54 @@ impl Party {
...
+            return Err(SecretError::BadIds(bad_ids));
+        }
+
+        let mut bad_shares = Vec::new();
+        for (i, s) in shares.iter() {
+            let Ai = &A[*i];
+            if s * G != (0..Ai.A.len()).fold(Point::zero(), |s, j| s + (self.id() ^ j) * Ai.A[j]) {
+                bad_shares.push(*i);
+            }
+        }
```
Comment (netrome) [inline src/v1.rs:130]:
```suggestion
        let bad_shares: Vec<usize> = shares
            .iter()
            .filter(|(i, s)| {
                let Ai = &A[**i];
                *s * G == (0..Ai.A.len()).fold(Point::zero(), |s, j| s + (self.id() ^ j) * Ai.A[j])
            })
            .map(|(i, _)| *i)
            .collect();
```

File: src/v1.rs [hunk 3]
```diff
@@ -161,27 +194,35 @@ impl SignatureAggregator {
...
-    ) -> Signature {
+    ) -> Result<Signature, SigningError> {
         assert_eq!(nonces.len(), sig_shares.len());
 
         let signers: Vec<usize> = sig_shares.iter().map(|ss| ss.id).collect();
         let (R_vec, R) = compute::intermediate(msg, &signers, nonces);
         let mut z = Scalar::zero();
         let c = compute::challenge(&self.key, &R, msg);
-
+        let mut bad_party_sigs = Vec::new();
```
Comment (netrome) [inline src/v1.rs:204]: Similar to my earlier suggestions, I'd prefer to skip the mutation here and build the collection with iterators `(0..sig_shares.len()).filter(...).collect()`.

**Review by netrome — APPROVED** [review]
LGTM. This would be even better with `Display` implemented on the error types. Do as you please with my "iterator-collect" suggestions.

File: src/errors.rs [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+    MissingShares(Vec<usize>),
+    BadIds(Vec<usize>),
+    BadShares(Vec<usize>),
+}
+
+#[derive(Debug)]
+pub enum SigningError {
+    BadPartySigs(Vec<usize>),
+    BadGroupSig,
+}
```
[the remark this answers is not in the document]
Comment (xoloki) [reply src/errors.rs]: Done.

File: src/v1.rs (hunk 1 above)
Comment (xoloki) [reply src/v1.rs]: Done.

**Pull request merged**